daras_ai_v2/search_ref.py

```diff
@@ -302,3 +305,47 @@ def render_output_with_refs(state, height=500):
 def generate_footnote_symbol(idx: int) -> str:
     quotient, remainder = divmod(idx, len(FOOTNOTE_SYMBOLS))
     return FOOTNOTE_SYMBOLS[remainder] * (quotient + 1)
+
+
+#
+def generate_text_fragment(text, min_len=0, max_len=30):
+    """
+    Generate a text fragment for a given text.
+
+    Args:
+        text (str): The input text to generate a fragment from.
+        min_len (int): The minimum length for the prefix and suffix.
+        max_len (int): The maximum length for the prefix and suffix.
+
+    Returns:
+        str: The encoded text fragment.
+    """
+    def find_alphanumeric_window(s, window_size, reverse=False):
+        range_func = range(len(s) - window_size + 1, 0, -1) if reverse else range(len(s) - window_size + 1)
+        for i in range_func:
+            window = s[i:i + window_size]
+            # Skip windows containing newline characters
+            if '\n' in window:
+                continue
+            # Updated regex to include allowed symbols and exclude '[', ']', and '.'
+            if re.match(r'^[a-zA-Z0-9\s"\'.,?:;()&^*$@!#]+$', window) and not re.search(r'[\[\].]', window):
+                words = window.strip().split()
+                if words and re.match(r'^[a-zA-Z0-9]+$', words[0]) and re.match(r'^[a-zA-Z0-9]+$', words[-1]):
+                    return window.strip()
+        # Fallback if no suitable window is found
+        return re.sub(r'[\[\].]', '', s[:window_size]).strip()
+
+    prefix = find_alphanumeric_window(text, max_len)
+    suffix = find_alphanumeric_window(text, max_len, reverse=True)
+
+    # Ensure prefix and suffix meet minimum length requirements
+    if len(prefix) < min_len:
+        prefix = find_alphanumeric_window(text, min_len)
+    if len(suffix) < min_len:
+        suffix = find_alphanumeric_window(text, min_len, reverse=True)
+    prefix = ' '.join(prefix.split()[:-1])
+    suffix = ' '.join(suffix.split()[1:])
+    # Assemble and encode the fragment
+    text_fragment = f"#:~:text={quote(prefix, safe='')},{quote(suffix, safe='')}"
```

Comment:
please use furl for url consturction as discussed earlier

Reply:
furl automatically encodes the `:~:` which should remain unencoded , couldn't find any work around except building url manually

Reply:
Oops

Reply:
Lets do this then, just in case the URL has a fragment already - 

```
return str(furl(url).remove(fragment=True)) + "#:~:text=..."
```

Reply:
It would break the pdf fragment for #page= 
we can do this selectively -
 
```
    url_obj = furl(url)
    if not url_obj.path.segments[-1].endswith('.pdf'):
        url_obj.remove(fragment=True)
    
    return str(url_obj) + text_fragment
```

Reply:
I believe for PDF we shouldn't do text fragment at all given it doesn't work. Plus, instead of checking for .pdf, we should do the inverse and check for html content type in `EmbeddingsReference.embedded_file.metadata.mime_type` here so we don't give the text fragment in non html pages (like slides docx etc)

https://github.com/GooeyAI/gooey-server/blob/51407ec9e94c4838e1036d010209e880f9e368a1/daras_ai_v2/vector_search.py#L224